facade/pendingdeployment.go

```diff
@@ -0,0 +1,122 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package facade
+
+import (
+	"sync"
+
+	"github.com/control-center/serviced/utils"
+)
+
+// Pending deployment represents a template deployment which is has been initiated
+// but has not yet completed.  It is synchronized, allowing access from multiple
+// goroutines.  In particular clients can register for asynchronous notification of
+// changes to its "status".
+type PendingDeployment struct {
+	status       utils.ValueChangePublisher
+	templateID   string
+	poolID       string
+	deploymentID string
+	templateName string
+	mutex        sync.RWMutex
+}
+
+// NewPendingDeployment returns a new PendingDeployment object.
+func NewPendingDeployment(deploymentID, templateID, poolID string) PendingDeployment {
+	return PendingDeployment{
+		status:       utils.NewValueChangePublisher("Starting"),
+		templateID:   templateID,
+		deploymentID: deploymentID,
+		poolID:       poolID,
+		mutex:        sync.RWMutex{},
+	}
+}
+
+// UpdateStatus updates the status of a PendingDeployment.
+func (d *PendingDeployment) UpdateStatus(status string) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	d.status.Set(status)
+}
+
+// SetTemplateName sets the template name of a PendingDeployment.
+func (d *PendingDeployment) SetTemplateName(name string) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	d.templateName = name
+}
+
+// GetStatus returns the current status of a PendingDeployment and a
+// channel which will be closewhen the status changes.
+func (d *PendingDeployment) GetStatus() (string, <-chan interface{}) {
+	// mutex lock not necessary; d.status.Get is protected
+	status, notify := d.status.Get()
+	return status.(string), notify
+}
+
+// GetInfo returns all information about a pending deployment in the
+// form of a map, appropriate for json serialization.
+func (d *PendingDeployment) GetInfo() map[string]string {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	status, _ := d.status.Get()
+	return map[string]string{
+		"TemplateID":   d.templateID,
+		"templateName": d.templateName,
+		"PoolID":       d.poolID,
+		"DeploymentID": d.deploymentID,
+		"status":       status.(string),
+	}
+}
+
+// PendingDeploymentMGr maintains a map from deployment ID to PendingDeployment
+// objects.
+type PendingDeploymentMgr struct {
+	deployments map[string]*PendingDeployment
+	mutex       sync.RWMutex
+}
+
+// NewPendingDeploymentMgr returns a new PendingDeploymentMgr
+func NewPendingDeploymentMgr() *PendingDeploymentMgr {
+	return &PendingDeploymentMgr{
+		deployments: make(map[string]*PendingDeployment),
+		mutex:       sync.RWMutex{},
+	}
+}
+
+// NewPendingDeployment allocates a new PendingDeployment with the given attributes
+// and associates it with the PendingDeploymentMgr
+func (dm *PendingDeploymentMgr) NewPendingDeployment(deploymentID, templateID, poolID string) *PendingDeployment {
+	deployment := NewPendingDeployment(deploymentID, templateID, poolID)
```

Comment:
Per your suggestion, I made NewPendingDeployment return an error in the case of a duplicate template id